source/CantorSchroederBernstein/CSB-WLPO.lagda

```diff
@@ -0,0 +1,286 @@
+ where
+
+ compute-construction-on-perfect-image :
+  (y : Y) →
+  (γ : is-perfect-image f g (g y)) →
+  (d : is-decidable (is-perfect-image f g (g y))) →
+  map-construction (g y) d ＝ y
+ compute-construction-on-perfect-image y γ (inl v') =
+  inverse-on-perfect-image-is-retraction lc-g y v'
+ compute-construction-on-perfect-image y γ (inr v) = 𝟘-elim (v γ)
+
+ compute-construction-on-not-perfect-image :
+  (y : Y) →
+  (nγ : ¬ is-perfect-image f g (g y)) →
+  (d :
+    is-decidable
+      ( is-perfect-image f g
+       ( element-in-nonperfect-fiber-of-not-perfect-image' αf ¬¬elim-g lc-g y nγ))) →
+  map-construction
+    (element-in-nonperfect-fiber-of-not-perfect-image' αf ¬¬elim-g lc-g y nγ)
+    (d) ＝
+  y
+ compute-construction-on-not-perfect-image y nγ (inl v) =
+  𝟘-elim (nonperfect-fiber-of-not-perfect-image-is-not-perfect' αf ¬¬elim-g lc-g y nγ v)
+ compute-construction-on-not-perfect-image y nγ (inr _) =
+  compute-element-in-nonperfect-fiber-of-not-perfect-image' αf ¬¬elim-g lc-g y nγ
+\end{code}
+
+The construction is an equivalence.
+
+\begin{code}
+ inverse-construction :
+   (y : Y) → is-decidable (is-perfect-image f g (g y)) → X
+ inverse-construction y (inl _) = g y
+ inverse-construction y (inr nγ) =
+   element-in-nonperfect-fiber-of-not-perfect-image' αf ¬¬elim-g lc-g y nγ
+
+ construction-is-retraction :
+  (y : Y)
+  (d : is-decidable (is-perfect-image f g (g y))) →
+  (d' : is-decidable (is-perfect-image f g (inverse-construction y d))) →
+  map-construction (inverse-construction y d) d' ＝ y
+ construction-is-retraction y (inl γ) =
+   compute-construction-on-perfect-image y γ
+ construction-is-retraction y (inr nγ) =
+   compute-construction-on-not-perfect-image y nγ
+
+ inverse-construction-CSB :
+  ((y : Y) → is-decidable (is-perfect-image f g (g y))) → Y → X
+ inverse-construction-CSB D y = inverse-construction y (D y)
+
+ is-section-inverse-construction-CSB :
+  (D : (x : X) → is-decidable (is-perfect-image f g x)) →
+  map-construction-CSB D ∘ inverse-construction-CSB (D ∘ g) ∼ id
+ is-section-inverse-construction-CSB D y =
+   construction-is-retraction y
+    (D (g y))
+    (D (inverse-construction-CSB (D ∘ g) y))
+
+ map-construction-CSB-has-section
+  : (D : (x : X) → is-decidable (is-perfect-image f g x))
+  → has-section (map-construction-CSB D)
+ map-construction-CSB-has-section D =
+  ( inverse-construction-CSB (D ∘ g) ,
+    is-section-inverse-construction-CSB D)
+
+ retract-CSB :
+  ((x : X) → is-decidable (is-perfect-image f g x)) → retract Y of X
+ retract-CSB D =
+  ( map-construction-CSB D , map-construction-CSB-has-section D)
+
+ construction-CSB-is-equiv :
+  left-cancellable f →
+  (D : (x : X) → is-decidable (is-perfect-image f g x))
+  → is-equiv (map-construction-CSB D)
+ construction-CSB-is-equiv lc-f D =
+  lc-retractions-are-equivs (map-construction-CSB D)
+   (map-construction-CSB-is-left-cancellable lc-f D)
+   (map-construction-CSB-has-section D)
+
+ equiv-CSB :
+  left-cancellable f
```

Comment:
The ":" should start in the next line (everywhere), unless you indent everything after the colon. The arrows should be indendent in the same column as the colon (in all cases).